examples/sdk-config.yaml

```
@@ -0,0 +1,206 @@
+# file configuration.
+#
+# Because env vars are ignored except for env var substitution, if "disabled: ${OTEL_SDK_DISABLED:-false}"
+# is replaced with "disabled: false", setting OTEL_SDK_DISALBED env var will not have any effect.
+# See https://opentelemetry.io/docs/specs/otel/configuration/file-configuration/ for more information.
+# The following spec defined env vars are NOT referenced and are thus ignored:
+#
+# - OTEL_RESOURCE_ATTRIBUTES
+# - OTEL_LOG_LEVEL
+# - OTEL_PROPAGATORS
+# - OTEL_TRACES_SAMPLER
+# - OTEL_TRACES_SAMPLER_ARG
+# - OTEL_EXPORTER_ZIPKIN_ENDPOINT
+# - OTEL_EXPORTER_ZIPKIN_TIMEOUT
+# - OTEL_EXPORTER_PROMETHEUS_HOST
+# - OTEL_EXPORTER_PROMETHEUS_PORT
+# - OTEL_TRACES_EXPORTER
+# - OTEL_METRICS_EXPORTER
+# - OTEL_LOGS_EXPORTER
+# - OTEL_METRICS_EXEMPLAR_FILTER
+# - OTEL_EXPORTER_OTLP_{SIGNAL}_ENDPOINT
+# - OTEL_EXPORTER_OTLP_{SIGNAL}_INSECURE
+# - OTEL_EXPORTER_OTLP_{SIGNAL}_CERTIFICATE
+# - OTEL_EXPORTER_OTLP_{SIGNAL}_CLIENT_KEY
+# - OTEL_EXPORTER_OTLP_{SIGNAL}_CLIENT_CERTIFICATE
+# - OTEL_EXPORTER_OTLP_HEADERS, OTEL_EXPORTER_OTLP_{SIGNAL}_HEADERS
+# - OTEL_EXPORTER_OTLP_{SIGNAL}_COMPRESSION
+# - OTEL_EXPORTER_OTLP_{SIGNAL}_TIMEOUT
+# - OTEL_EXPORTER_OTLP_{SIGNAL}_PROTOCOL
+
+# The file format version
+file_format: "0.1"
+
+# Configure if the SDK is disabled or not.
+disabled: ${OTEL_SDK_DISABLED:-false}
+
+# Configure resource for all signals.
+resource:
+  # Configure resource attributes.
+  attributes:
+    # Configure `service.name` resource attribute
+    service.name: ${OTEL_SERVICE_NAME:-unknown_service}
+
+# Configure general attribute limits. See also tracer_provider.limits, logger_provider.limits.
+attribute_limits:
+  # Configure max attribute value size.
+  attribute_value_length_limit: ${OTEL_ATTRIBUTE_VALUE_LENGTH_LIMIT}
+  # Configure max attribute count.
+  attribute_count_limit: ${OTEL_ATTRIBUTE_COUNT_LIMIT:-128}
+
+# Configure text map context propagators.
+propagator:
+  composite: [tracecontext, baggage]
+
+# Configure tracer provider.
+tracer_provider:
+  # Configure span processors.
+  processors:
+    # Configure a batch span processor.
+    - batch:
+        # Configure delay interval (in milliseconds) between two consecutive exports.
+        schedule_delay: ${OTEL_BSP_SCHEDULE_DELAY:-5000}
+        # Configure maximum allowed time (in milliseconds) to export data.
+        export_timeout: ${OTEL_BSP_EXPORT_TIMEOUT:-30000}
+        # Configure maximum queue size.
+        max_queue_size: ${OTEL_BSP_MAX_QUEUE_SIZE:-2048}
+        # Configure maximum batch size.
+        max_export_batch_size: ${OTEL_BSP_MAX_EXPORT_BATCH_SIZE:-512}
+        # Configure exporter.
+        exporter:
+          # Configure exporter to be OTLP.
+          otlp:
+            # Configure protocol.
+            protocol: ${OTEL_EXPORTER_OTLP_PROTOCOL:-http/protobuf}
+            # Configure endpoint.
+            endpoint: ${OTEL_EXPORTER_OTLP_PROTOCOL:-http://localhost:-4318}
```

```suggestion
            endpoint: ${OTEL_EXPORTER_OTLP_PROTOCOL:-http://localhost:4318/v1/traces}
```

---

That would make it `OTEL_EXPORTER_OTLP_TRACES_ENDPOINT`

Also, its currently using `OTEL_EXPORTER_OTLP_PROTOCOL` when I'm guessing it wants either `OTEL_EXPORTER_OTLP_TRACES_ENDPOINT` or to actually put the path outside the environment variable default to apply to both:

```
endpoint: ${OTEL_EXPORTER_OTLP_ENDPOINT:-http://localhost:4318}/v1/traces
```

---

Good point on the TRACES part. So, something like this then:

```suggestion
            endpoint: ${OTEL_EXPORTER_OTLP_TRACES_ENDPOINT:-http://localhost:4318/v1/traces}
```

This also means using the TRACES env var for PROTOCOL as well.

---

We have to decide between the generic version of these properties and the signal specific. I think the generic version of these is likely to be more popular. But the generic version of the property [doesn't require you to configure](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/exporter.md#endpoint-urls-for-otlphttp) a path. 

One thing which is still ambiguous is whether the endpoints specified in file config have paths appended automatically if the protocol is `http/protobuf` and the path is omitted. I lean towards yes.

---

I'm ok with going with the generic endpoint as an example for now

---

In my understanding, the only viable setting is:

```
endpoint: ${OTEL_EXPORTER_OTLP_TRACES_ENDPOINT:-http://localhost:4318/v1/traces}
```

If any of the following are used instead:

```
endpoint: ${OTEL_EXPORTER_OTLP_TRACES_ENDPOINT:-http://localhost:4318/}
```

```
endpoint: ${OTEL_EXPORTER_OTLP_ENDPOINT:-http://localhost:4318/}
```

then please point to the spec that mandates to add `v1/traces` in the path, and I don't mean the definition of existing pre config.yaml environment variable `OTEL_EXPORTER_OTLP_ENDPOINT` here, because this must also apply to any variable, including `MY_ENDPOINT`.

---

I think jack was saying it should just be `endpoint: ${OTEL_EXPORTER_OTLP_ENDPOINT:-http://localhost:4318}`. Which is correct. `v1/traces` is added by default when using the non-signal specific endpoint configuration.

I was only suggesting the signal specific var because you were suggesting the addition of `v1/traces`.

---

To clarify, this is how opentelemetry-cpp parses environment variables (before config.yaml)

```
std::string GetOtlpDefaultHttpTracesEndpoint()
{
  constexpr char kSignalEnv[]  = "OTEL_EXPORTER_OTLP_TRACES_ENDPOINT";
  constexpr char kGenericEnv[] = "OTEL_EXPORTER_OTLP_ENDPOINT";
  constexpr char kDefault[]    = "http://localhost:4318/v1/traces";

  std::string value;
  bool exists;

  exists = sdk_common::GetStringEnvironmentVariable(kSignalEnv, value);
  if (exists)
  {
    return value;
  }

  exists = sdk_common::GetStringEnvironmentVariable(kGenericEnv, value);
  if (exists)
  {
    value += "/v1/traces";
    return value;
  }

  return kDefault;
}
```

The `v1/traces` is added when the variable name is `OTEL_EXPORTER_OTLP_ENDPOINT`, it is not added later in the SDK when a trace exporter needs an endpoint.

---

Argh, sorry, I misunderstood the context from the beginning. This is `endpoint` within the trace provider. So you are right, it should be `endpoint: ${OTEL_EXPORTER_OTLP_ENDPOINT:-http://localhost:4318}/v1/traces`

---

Means the example config won't support setting the path, but it can't support everything?

---

I think there is more to it.

Options are correlated, the path is only used for OTLP HTTP but not for GRPC if I recall correctly, so something like:

```
endpoint: ${OTEL_EXPORTER_OTLP_ENDPOINT:-http://localhost:4318}/v1/traces
```

really assumes OTLP HTTP.

Where:

```
endpoint: ${OTEL_EXPORTER_OTLP_TRACES_ENDPOINT:-http://localhost:4318/v1/traces}
```

can be replaced by the GRPC endpoint (using port 4317 while at it)


The whole point of environment variables is to change parts while not having to touch the config.yaml file.

If the endpoint changes between OTLP HTTP and OTLP GRPC, I don't see how OTEL_EXPORTER_OTLP_ENDPOINT can be used anywhere.

---

So summarize my comments, the content of the config.yaml as written in this PR:

```
            # Configure protocol.
            protocol: ${OTEL_EXPORTER_OTLP_PROTOCOL:-http/protobuf}
            # Configure endpoint.
            endpoint: ${OTEL_EXPORTER_OTLP_PROTOCOL:-http://localhost:-4318}
```

does not allow to connect using the OTLP HTTP protocol to an opentelemetry-collector.

---

Right, it should definitely not be that because it will mean the trace exporter won't send to `v1/traces`.

---

These is a good convo. Thanks for calling this out @marcalff. 

>> v1/traces is added by default when using the non-signal specific endpoint configuration
> Where, by whom ?

When the `OTEL_EXPORTER_OTLP_ENDPOINT` env var is set, the "where" & "whom" is whatever component is interpreting the env vars and setting up the exporter. According this [this](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/exporter.md#endpoint-urls-for-otlphttp) (as you note) the code is responsible for doing different things with the endpoint value based on whether the signal-specific or generic env var is used, and based on whether the protocol is `grpc` or `http/protobuf`. 

You've suggested the following alternatives:

```
endpoint: ${OTEL_EXPORTER_OTLP_TRACES_ENDPOINT:-http://localhost:4318/v1/traces}
endpoint: ${OTEL_EXPORTER_OTLP_ENDPOINT:-http://localhost:4318}/v1/traces
```

The problem with both of these (and not necessarily a dealbreaker) is that if the sets `OTEL_EXPORTER_OTLP_PROTOCOL=grpc`, then the endpoint breaks. They need to configure `OTEL_EXPORTER_OTLP_ENDPOINT` / `OTEL_EXPORTER_OTLP_TRACES_ENDPOINT` in lockstep with `OTEL_EXPORTER_OTLP_PROTOCOL`.

Another alternative is to say that code interpreting the `*.otlp.protocol` and `*.otlp.endpoint` values of the model has additional responsibilities besides plugging these raw values into the exporter. Essentially carrying over the logic interpreting `OTEL_EXPORTER_OTLP_ENDPOINT` to comments:
```
tracer_provider:
  processors:
    - batch:
       exporter:
         otlp:
           # Configure protocol.
           protocol: ${OTEL_EXPORTER_OTLP_PROTOCOL:-http/protobuf}
           # Configure endpoint.
           #
           # When .protocol is set to http/protobuf and the endpoint URL contains no path element, implementations MUST set the path to be "/v1/traces".
           endpoint: ${OTEL_EXPORTER_OTLP_ENDPOINT:-http://localhost:4318}
```

The problem with this is that it makes it impossible to specify that data should be sent do an endpoint without any path at all, and the env var spec seems to have explicitly [carved out an exception for this](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/exporter.md#endpoint-urls-for-otlphttp):

> For the per-signal variables (OTEL_EXPORTER_OTLP_<signal>_ENDPOINT), the URL MUST be used as-is without any modification. The only exception is that if an URL contains no path part, the root path / MUST be used (see [Example 2](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/exporter.md#example-2)).


My conclusion is that @marcalff is correct and the only correct alternatives are:

```
endpoint: ${OTEL_EXPORTER_OTLP_TRACES_ENDPOINT:-http://localhost:4318/v1/traces}
endpoint: ${OTEL_EXPORTER_OTLP_ENDPOINT:-http://localhost:4318}/v1/traces
```

And I'm now leaning towards the first options because the idea of appending a path to the `OTEL_EXPORTER_OTLP_ENDPOINT` env var is a code smell to me.
```
endpoint: ${OTEL_EXPORTER_OTLP_TRACES_ENDPOINT:-http://localhost:4318/v1/traces}
```

What do you all think?

---

Agreed we can't require the code interpreting `endpoint` to append a path. It wouldn't just mean you can't have an empty path, but it would mean you can't have any alternative path at all because it would always append `v1/traces`. There are already vendors using different paths.

I also don't think the second option is a code smell *but* I do agree that because it can't take into account the protocol that the first option is the only real answer.

---

> There are already vendors using different paths.

I can confirm this.

See for example:

https://docs.oracle.com/en-us/iaas/application-performance-monitoring/doc/configure-open-source-tracing-systems.html#GUID-4D941163-F357-4839-8B06-688876D4C61F


    For traces authenticated with public data key, add the following:

    https://<dataUploadEndpoint>/20200101/opentelemetry/public/v1/traces
    For traces authenticated with private data key, add the following:

    https://<dataUploadEndpoint>/20200101/opentelemetry/private/v1/traces

---

Those still end in `v1/traces`. So having that appended to each would still work fine. I believe some vendors don't have the `v1/traces` at the end.